File: C/Types/Struct/struct.c
#include <stdio.h>
#include <mpi.h>
#include <string.h>


//My struct. I only want to send a, b and c, NOT inter
typedef struct {
    int a;
    char inter;
    int b[5];
    float c[6];
} mystruct;

int main(int argc, char** argv)
{

  int rank, i, off;
  int lengths[3];
  MPI_Aint displacements[3];
  MPI_Datatype types[3];
  MPI_Datatype struct_type;
  mystruct data;

  char buffer[400];

  MPI_Init(&argc, &argv);
  MPI_Comm_rank(MPI_COMM_WORLD, &rank);

  // Be careful of using "offsetof" in C++ since it only works with standard
  // layout types before C++17 and is "conditionally-supported" for types that
  // are not standard layout types from C++17. 
  // See https://en.cppreference.com/w/cpp/named_req/StandardLayoutType
  // and https://en.cppreference.com/w/cpp/types/is_standard_layout
  // In C requires C89 (ANSI C) or newer
  displacements[0] = offsetof(mystruct, a);
  displacements[1] = offsetof(mystruct, b);
  displacements[2] = offsetof(mystruct, c);

  types[0] = MPI_INT; types[1] = MPI_INT; types[2] = MPI_FLOAT;
  //Lengths are in multiples of the associated type
  lengths[0] = 1; lengths[1] = 5; lengths[2] = 6;

  if (rank == 0) {
    printf("Byte offsets are %li %li %li\n", displacements[0], 
        displacements[1], displacements[2]);
  }

  //On rank 0 set some dummy values in my struct
  //Don't do this on other ranks so their version
  //Will be in a different and random state
  if (rank == 0){
    data.a = 100;
    for (i = 0; i<5; ++i){
      data.b[i] = i;
    }
    for (i = 0; i<6; ++i){
      data.c[i] = 100.0/(float)i;
    }
  }

  data.inter = rank;

  //Create and commit the type
  MPI_Type_create_struct(3, lengths, displacements, types, &struct_type);
  MPI_Type_commit(&struct_type);

  //Use MPI_Bcast to send the value from rank 0 to all other ranks
  MPI_Bcast(&data, 1, struct_type, 0, MPI_COMM_WORLD);

  //Output to screen. Write all output to a single string first and then print
  //to maximise the chance that it all gets printed in one chunk
  memset(buffer, 0, 400);
  off=0;
  off+=sprintf(buffer,"============\n");
  off+=sprintf(buffer+off,"rank: %i\n", rank);
  off+=sprintf(buffer+off,"a   : %i\n", data.a);
  off+=sprintf(buffer+off,"inter   : %i\n", data.inter);
  off+=sprintf(buffer+off,"NOTE THAT INTER IS DIFFERENT!\n");
  off+=sprintf(buffer+off,"b   : ");
  for (i = 0; i<5; ++i){
    off+=sprintf(buffer+off,"%i ", data.b[i]);
  }
  off+=sprintf(buffer+off,"\nc   :");
  for (i = 0; i<6; ++i){
    off+=sprintf(buffer+off,"%.6f ", data.c[i]);
  }
  off+=sprintf(buffer+off,"\n");
  printf("%s", buffer);

  //Free the type and Finalize MPI
  MPI_Type_free(&struct_type);
  MPI_Finalize();
}

File: Fortran/Types/Struct/struct.f90
PROGRAM type

  USE mpi
  IMPLICIT NONE

  !Only want to send a,b and c, not inter. Could send all of them
  !but demonstrate that you don't have to
  TYPE mytype
    !THIS IS VERY IMPORTANT! The Fortran standard allows reordering of elements
    !in types unless you either specify SEQUENCE or BIND(C) attributes
    SEQUENCE
    INTEGER :: a
    CHARACTER(LEN=1) :: inter
    INTEGER, DIMENSION(5) :: b
    REAL(SELECTED_REAL_KIND(6, 37)), DIMENSION(6) :: c
  END TYPE mytype

!This is the alternative using BIND(C)
!  TYPE, BIND(C) :: mytype
!    INTEGER :: a
!    CHARACTER(LEN=1) :: inter
!    INTEGER, DIMENSION(5) :: b
!    REAL(SELECTED_REAL_KIND(6, 37)), DIMENSION(6) :: c
!  END TYPE mytype

  INTEGER :: rank
  INTEGER :: ierr, i
  INTEGER(MPI_ADDRESS_KIND) :: base
  INTEGER(MPI_ADDRESS_KIND), DIMENSION(3) :: displacements
  INTEGER, DIMENSION(3) :: lengths, types
  INTEGER :: struct_type
  TYPE(mytype) :: data

  CALL MPI_Init(ierr)

  CALL MPI_Comm_rank(MPI_COMM_WORLD, rank, ierr)

  CALL MPI_Get_address(data, base, ierr)
  CALL MPI_Get_address(data%a, displacements(1), ierr)
  CALL MPI_Get_address(data%b, displacements(2), ierr)
  CALL MPI_Get_address(data%c, displacements(3), ierr)
  !MPI_Get_address gets the absolute address from MPI_BOTTOM so you have to
  !subtract the location of the base of your type from the displacements of
  !the elements
  displacements(1) = MPI_Aint_diff(displacements(1), base)
  displacements(2) = MPI_Aint_diff(displacements(2), base)
  displacements(3) = MPI_Aint_diff(displacements(3), base)

  IF (rank == 0) THEN
    PRINT "(A, I4, I4, I4)", "Byte offsets are ", displacements
  END IF

  types = [MPI_INTEGER, MPI_INTEGER, MPI_REAL]
  !Lengths are in multiples of the associated type
  lengths = [1, 5, 6]
  
  !Create the type
  CALL MPI_Type_create_struct(3, lengths, displacements, types, struct_type, &
      ierr)
  !Register the type
  CALL MPI_Type_commit(struct_type, ierr)

  !On rank 0 set some dummy values in my type
  !Don't do this on other ranks so their version
  !Will be in a different and random state
  IF (rank == 0) THEN
    data%a = 100
    DO i = 1, 5
      data%b(i) = i
    END DO
    DO i = 1, 6
      data%c(i) = 100.0/REAL(i, SELECTED_REAL_KIND(6, 37))
    END DO
  END IF

  !Create a character starting at A and increasing for each rank. Cycles at 26
  data%inter = ACHAR(ICHAR('A') + MOD(rank,26))

  IF (rank == 0) THEN
    PRINT *, 'To read the output from this example, open the files fort.1**'
    PRINT *, 'The output from rank 0 will be in fort.100'
    PRINT *, 'The output from rank 1 will be in fort.101 etc.'
  END IF

  CALL MPI_Bcast(data, 1, struct_type, 0, MPI_COMM_WORLD, ierr)

  WRITE(100+rank,*) 'A     : ', data%a
  WRITE(100+rank,*) 'Inter : ', data%inter
  WRITE(100+rank,*) 'NOTE THAT INTER IS DIFFERENT!'
  WRITE(100+rank,*) 'B     : ', data%b
  WRITE(100+rank,*) 'C     : ', data%c

  CALL MPI_Type_free(struct_type, ierr)
  CALL MPI_Finalize(ierr)

END PROGRAM type
